src/core/DY.Core.Label.fst

```diff
@@ -318,3 +292,92 @@ let join_flow_to_public_eq tr x1 x2 =
   reveal_opaque (`%can_flow) (can_flow);
   reveal_opaque (`%join) (join);
   reveal_opaque (`%public) (public)
+
+/// Generic label for states:
+/// `state_pred_label p` is corrupt when
+/// there exists a corrupt `SetState` event satisfying the predicate `p`.
+/// It can for example be used to depict any state of a principal
+/// (e.g. as done by `principal_label`)
+
+[@@"opaque_to_smt"]
+val state_pred_label:
+  (principal -> state_id -> bytes -> prop) ->
+  label
+let state_pred_label p = mk_label {
+  is_corrupt = (fun tr ->
+    exists prin sess_id content.
+      was_corrupt tr prin sess_id content /\
+      p prin sess_id content
+  );
+  is_corrupt_later = (fun tr1 tr2 -> ());
+}
+
+val state_pred_label_pred_can_flow:
+  (principal -> state_id -> bytes -> prop) ->
+  (principal -> state_id -> bytes -> prop) ->
+  prop
+let state_pred_label_pred_can_flow p1 p2 =
+  forall p s c. p2 p s c ==> p1 p s c
+
+val state_pred_label_can_flow_state_pred_label:
+  tr:trace ->
+  p1:(principal -> state_id -> bytes -> prop) -> p2:(principal -> state_id -> bytes -> prop) ->
+  Lemma
+  (requires state_pred_label_pred_can_flow p1 p2)
+  (ensures state_pred_label p1 `can_flow tr` state_pred_label p2)
+  [SMTPat (state_pred_label p1 `can_flow tr` state_pred_label p2)]
+let state_pred_label_can_flow_state_pred_label tr p1 p2 =
+  reveal_opaque (`%is_corrupt) is_corrupt;
+  reveal_opaque (`%can_flow) (can_flow);
+  reveal_opaque (`%state_pred_label) (state_pred_label)
+
+val state_pred_label_can_flow_public:
+  tr:trace ->
+  p:(principal -> state_id -> bytes -> prop) ->
+  Lemma (
+    (state_pred_label p) `can_flow tr` public
+    <==> (
+      exists prin sess_id content.
+        was_corrupt tr prin sess_id content /\
+        p prin sess_id content
+    )
+  )
+  [SMTPat ((state_pred_label p) `can_flow tr` public)]
+let state_pred_label_can_flow_public tr p =
+  reveal_opaque (`%is_corrupt) is_corrupt;
+  reveal_opaque (`%state_pred_label) (state_pred_label);
+  FStar.Classical.forall_intro (FStar.Classical.move_requires (event_exists_fmap_trace_eq forget_label tr));
+  FStar.Classical.forall_intro_2 (FStar.Classical.move_requires_2 (event_at_fmap_trace_eq forget_label tr))
+
+val principal_label_pred:
+  principal ->
+  principal -> state_id -> bytes -> prop
+let principal_label_pred prin1 prin2 _ _ =
```

Comment:
Why do we have the `state_id` and `bytes` arguments, if we don't use them?
(Same for `principal_state_label_pred`)

Reply:
This function should be read as taking a `principal` as input, and outputting a `principal -> state_id -> bytes -> prop`, which is the input of `state_pred_label`.
I can add parenthesis to make that more clear, or define a type alias for `principal -> state_id -> bytes -> prop`

Reply:
Another possibility is to write
```fstar
let principal_label_pred prin1 =
  fun prin2 _ _ -> ...
```
(which complements nicely adding parenthesis)

Reply:
Ah I see. Can we do all of your suggestions?
I think. having a type alias for the input of `state_pred_label` and defining the function on the right side of the let would make the connection to `state_pred_label` more apparent.

Reply:
Done. I named it `state_pred_label_input`, as it is the input to `state_pred_label`. I initially went to `state_pred_label_pred` but the "pred pred" is confusing I think.